### ISSUE TITLE
Implicit Subscription Dates

### ISSUE THREAD

**Issue by pascalfree** [juntagrico/juntagrico#366]
# Motivation
Currently activation, cancellation and deactivation dates must be set for subscriptions as well as subscription parts. The validities of these dates depend on each other requiring a lot of checks which cause some inconviniences and confusion in handling the subscriptions e.g. #355 

# Idea

* In theory the activation date of the subscription must always match the activation date of the first activated part. Otherwise the subscription would have been active without any active parts, which is prohibited. An active part in a waiting subscription is equally illegal.
* The deactivation date of the subscription must match the deactivation date of the last deactivated part, with the same reasoning as above
* The cancellation date of the subscription must be empty, if not all parts are cancelled. if all parts are cancelled the cancellation date of the subscription should logically match the date of the last cancelled part

In conclusion these dates of the subscription are redundant and can be derived directly from the parts of the subscription.

# Approaches

1. Easy: The fields for the dates would stay in place, but would be made uneditable by the managers. Whenever a part is saved the dates of the subscription would be updated automatically. activating the subscription is equivalent to activating all of its parts, and same for cancelling and deactivating.
2. Harder: The fields are replaced by properties. Some queries might have to be modified in this case and would become more complex.

**Comment by motu55** [comment]
some quick unsorted thoughts on that issue

an issue which is not covered by this  approach is that a subscription could be reactivated by adding an active subscription part
or if a waiting subscription part is added to a deactivated subscription all havoc breakes loose
probably this could be prevented by disabling the add functionality in the inlines.

also the coupling between subscription membership dates and part dates would become harder.

also what happens when a part is added with an earlier activation date than the others? this might be an error or it might be correct; how will we detect that?. by forcing the admin to also alter the corresponding subscription date we can detect that.

It seems to me that we do not throw a way our many checks, but we would be replacing them with a lot of different ones, and would end up with less functionality and way more complex queries than we already have.

Also the creation, activation, cancelation and deactivation do usually not happen in the admin interface. admins only go there to fix stuff (which is currently possible, when alos annoying)

I think it is less effort to focus on #355 and find a good solution there instead of changing the whole model where we do not have a guarantee that thing will get better

**Comment by pascalfree** [comment]
What i missed to point out, is that approach 1 would already implement the entire idea. Approach 2 would go further and most likely make some things harder as you mentioned.

Regarding

- Reactivation: Why do we prohibit reactivation so forcefully? Is it a billing issue?
- Coupling to membership dates would be harder: Yes, but only approach 2
- Adding earlier activation: That would be a discussion of the advantages and disatvantages of redundancy (i.e. double-entry bookkeeping). One could argue that with the current model the activation date of the sub can be modified in such a way, that then there is a time period in the past where sub was active but without any active parts (which is inconsistent). To avoid date typos we could show a warning, if the new date is earlier than the current (calculated) activation date.
- Will make some checks harder: Yes for approach 2.
- Admin only to fix: Agreed, that's why I wouldn't put obstacles in the ways for chaning the dates in the admin interface.

I guess I'll start a poc some day to get a more concrete idea of the implications.

**Comment by motu55** [comment]
@samuel-iseli  we would need some information to progress the discussion here
does the billing module when it calculates the price of an subscription the activation and deactivation dates of a subscriprion?
and how is the price calculated? by iterationg over all the parts and calculate the price for that part in the given period and then summed up?

**Comment by mbschenkel** [comment]
My question / proposal goes maybe a bit far from the current state and the discussion above, and could mean considerable rework.  But still:  Is there a strong reason that the member-subscription relationship is many-to-1 and not many-to-many?

If it was many-to-many, we could simply cancel one and recreate a new one later, having clear start- and end-dates for billing.
I'm also wondering whether we would even still need subscription and subscription part - but can't judge since we only offer one product.

Also this would (obviously) allow a member to be part of 2 subscriptions for different products.
We even discussed such a use-case recently for bioco, where we wanted to capture a special contract condition that only affected 1 out of 2 members in the same subscription, but realized this is not possible today.

**Comment by motu55** [comment]
The subscription is a Container to group parts members and extra subscriptions.
In theory you are right we could handle each part or extra subscription as a subscription and allow members to be part of multiple subscriptions.
In practice it is not a good idea since it blows up the depot lists and makes calculation if enough shares are available a complete nightmare.
This also applies to just remove the many to one relationship.
So this is an idea which we will not consider further.

**Comment by samuel-iseli** [comment]
Billing currently looks for subscription parts via subscriptions that are active during the billing period (business-year). I wanted to search for subscription-parts directly but that seemed unreliable, at least some months ago.
The extrasubscriptions are collected separately, also by their activation/deactivation-dates. 
All items whose active interval (activation_date to deactivation_date) overlaps with the billing period are considered for billing.
They are billed if they are not already on a bill in the period. Code is in juntagrico-billing.utils.billing.get_billable_items.

I think it would be nice if we could simplify the state handling somehow. I spend quite a lot of time fighting against the integrity checks.

I would favour the proposal where we change the activation/deactivation dates on the subscription into properties which are calculated based on the parts and extrasubscriptions.
The billing logic could be changed to directly query the subscription-parts instead of the subscriptions. That would be even simpler than it is now.
